### PR TITLE
feat(SPV-1295): testabilities for contacts admin

### PR DESCRIPTION
**Pull request by dzolt-4chain**
# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/contact/contact_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	tests := map[string]struct {
+		contactPaymail   string
+		creatorPaymail   string
+		fullName         string
+		metadata         *engine.Metadata
+		expectedError    error
+		expectedStatus   engine.ContactStatus
+		expectedFullName string
+	}{
+		"Happy path without metadata": {
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:24]:
```suggestion
		"Create contact without metadata": {
```

File: engine/contact/contact_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			contactPaymail:   fixtures.RecipientExternal.DefaultPaymail(),
+			creatorPaymail:   fixtures.Sender.DefaultPaymail(),
+			fullName:         "John Doe",
+			metadata:         nil,
+			expectedError:    nil,
+			expectedStatus:   engine.ContactNotConfirmed,
+			expectedFullName: "John Doe",
+		},
+		"Happy path with metadata": {
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:33]:
```suggestion
		"Create contact with metadata": {
```

File: engine/contact/contact_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			//given:
+			given, then := testabilities.New(t)
+
+			service, cleanup := given.Engine()
+			defer cleanup()
+
+			//when:
+			res, err := service.AdminCreateContact(context.Background(), tt.contactPaymail, tt.creatorPaymail, tt.fullName, tt.metadata)
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:56]: I feel that is unit tests instead of `res` (and later `WithResponse`) we should use word `contact` or something because this isn't the endpoint test and is rather focused on a specific feature, then the API with some responses.

File: engine/contact/contact_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		defer cleanup()
+
+		//and:
+		given.PaymailClient().WillRespondOnCapability(paymail.BRFCPki).WithInternalServerError()
+
+		//when:
+		res, err := service.AdminCreateContact(context.Background(), fixtures.RecipientExternal.DefaultPaymail(), fixtures.Sender.DefaultPaymail(), "John Doe", nil)
+
+		//then:
+		then.ErrorIs(err, spverrors.ErrGettingPKIFailed).WithNilResponse(res)
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:79]: If you would go with structure more like this:
```suggestion
		then.ContactCreated(res).WithError(err).ThatIs(spverrors.ErrGettingPKIFailed)
```
then in `WithError` you could require `res` (contact) to be `nil`, as this is what we always expect when there is an error.

ℹ️ You can find example of something like that in [create_draft_test.go](https://github.com/bitcoin-sv/spv-wallet/blob/3312ad9f92c0acb11723db02e18850535b08b549/engine/transaction/outlines/create_draft_test.go#L51)

File: engine/contact/contact_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+			given, then := testabilities.New(t)
+
+			service, cleanup := given.Engine()
+			defer cleanup()
+
+			//when:
+			res, err := service.AdminCreateContact(context.Background(), tt.contactPaymail, tt.creatorPaymail, tt.fullName, tt.metadata)
+
+			//then:
+			then.NoError(err).WithResponse(res).WithStatus(tt.expectedStatus).WithFullName(tt.expectedFullName)
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:59]: IMHO you should also 
1. check the PubKey - if it matches the one from external paymail server. 
2. And maybe also check the Paymail field that it is paymail of contact
3. and maybe could add an assertion that contact was created for user, and pass the fixtures.Sender and check if OwnerXpubID is matching user.XPubID

to sum up, I would see it like this:
```suggestion
			then.ContactCreated(contact).WithoutError(err)
			  .ForUser(fixtures.Sender) // Maybe it could be named InContactBookOf(fixtures.Sender)
			  .WithContactPaymailOf(fixtures.RecipientExternal)
			  .WithPubKey(fixtures.RecipientExternalPKI)
			  .WithStatus(tt.expectedStatus)
			  .WithFullName(tt.expectedFullName)
```

alternatively it could be something like this:
```suggestion
			then.ContactCreated(contact).WithoutError(err)
			  .ForUser(fixtures.Sender)  // Maybe it could be named InContactBookOf(fixtures.Sender)
			  .ToCounterparty(fixtures.RecipientExternal) // and here we check the paymail and PubKey
			  .WithStatus(tt.expectedStatus)
			  .WithFullName(tt.expectedFullName)
```

File: engine/contact/contact_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+func Test_ClientService_AdminCreateContact_Fail(t *testing.T) {
+	tests := map[string]struct {
+		contactPaymail   string
+		creatorPaymail   string
+		fullName         string
+		expectedError    error
+		expectedStatus   engine.ContactStatus
+		expectedFullName string
+	}{
+		"Edge case: Creator paymail not found": {
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:92]: I would rather go with:
- `should fail when ...`
- `should return error when ...`
- `return error when ...`

instead of `Edge case: ` 

It's actually not even an edge case, it's rather error case 🙂

File: engine/contact/contact_test.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			expectedFullName: "",
+		},
+		"Edge case: missing creator paymail": {
+			contactPaymail: fixtures.RecipientExternal.DefaultPaymail(),
+			creatorPaymail: "",
+			fullName:       "John Doe",
+			expectedError:  spverrors.ErrMissingContactCreatorPaymail,
+		},
+		"Edge case: missing contact full name": {
+			contactPaymail: fixtures.RecipientExternal.DefaultPaymail(),
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:107]: there is no test for missing `contactPaymail`

File: engine/contact/testabilities/assert_contact.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	return a
+}
+
+func (a *assert) WithNilResponse(response *engine.Contact) ContactFailureAssertion {
+	a.require.Nil(response, "unexpected response")
+	return a
+}
+
+func (a *assert) WithStatus(status engine.ContactStatus) ContactSuccessAssertion {
+	a.require.Equal(status, a.response.Status, fmt.Sprintf("expected status: %s, actual: %s", status, a.response.Status))
```
Comment (dorzepowski) [inline engine/contact/testabilities/assert_contact.go:62]: There is no need for `fmt.Sprintf` because the methods in `require` nad `assert` are accepting `string` which is string template as the and the args to this template as the parameters

In other words, this should work:
```suggestion
	a.require.Equal(status, a.response.Status, "expected status: %s, actual: %s", status, a.response.Status)
```

Additionally maybe it's enough to just put string `"Contact has invalid status"`, because `require.Equal` will show the difference between expected and received values.

👇 same thing for the WithFullName

File: engine/contact/testabilities/assert_contact.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	return a
+}
+
+func (a *assert) WithFullName(fullName string) ContactSuccessAssertion {
+	a.require.Equal(fullName, a.response.FullName, fmt.Sprintf("expected fullName: %s, actual: %s", fullName, a.response.FullName))
```
Comment (dorzepowski) [inline engine/contact/testabilities/assert_contact.go:67]: In checks of the specific fields in contact (like WithFullName, WithStatus, etc.) I would prefer to use package `assert` instead of `require`.

File: engine/contact/testabilities/assert_contact.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	a.response = response
+	return a
+}
+
+func (a *assert) WithNilResponse(response *engine.Contact) ContactFailureAssertion {
+	a.require.Nil(response, "unexpected response")
+	return a
+}
+
+func (a *assert) WithStatus(status engine.ContactStatus) ContactSuccessAssertion {
```
Comment (dorzepowski) [inline engine/contact/testabilities/assert_contact.go:61]: because in those tests we're always expect to have status `ContactNotConfirmed` maybe it would be good to have (and call) method `NotConfirmed()` or `AsNotConfirmed()`.
so something like that:
```golang
func (a *assert) AsNotConfirmed() {
   return a.WithStatus(engine.ContactNotConfirmed)
}
```

File: engine/contact/contact_test.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+func Test_ClientService_AdminCreateContact_Success(t *testing.T) {
+	tests := map[string]struct {
+		contactPaymail   string
+		creatorPaymail   string
+		fullName         string
+		metadata         *engine.Metadata
+		expectedError    error
+		expectedStatus   engine.ContactStatus
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:21]: When created, we always expect the status to be `ContactNotConfirmed` so I believe that this field is not needed.

File: engine/contact/contact_test.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func Test_ClientService_AdminCreateContact_Fail(t *testing.T) {
+	tests := map[string]struct {
+		contactPaymail   string
+		creatorPaymail   string
+		fullName         string
+		expectedError    error
+		expectedStatus   engine.ContactStatus
+		expectedFullName string
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:90]: In error cases you don't need the expected status and full name as actually the contact is nil so it doesn't has those fields 🙂

File: engine/contact/contact_test.go
Comment (dorzepowski) [inline engine/contact/contact_test.go:1]: It would be good to have also endpoint tests (but it could be done in a separate PR)

File: engine/contact/contact_test.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	"testing"
+
+	"github.com/bitcoin-sv/go-paymail"
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/engine/contact/testabilities"
+	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
+	"github.com/bitcoin-sv/spv-wallet/engine/tester/fixtures"
+)
+
+func Test_ClientService_AdminCreateContact_Success(t *testing.T) {
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:14]: You should remove the equivalent tests from the engine/contact_service_test.go

**Review by dorzepowski — CHANGES_REQUESTED** [review]
Great job, 
For me, it reads much better then the previous approach.

File: engine/contact/testabilities/assert_contact.go [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+func (a *assert) NoError(err error) ContactSuccessAssertion {
+	a.require.NoError(err, "Record transaction outline has error")
+	return a
+}
+
+func (a *assert) ErrorIs(err, expectedError error) ContactFailureAssertion {
+	require.Error(a.t, err, "Record transaction outline has no error")
+	require.ErrorIs(a.t, err, expectedError, "Record transaction outline has wrong error")
+	return a
+}
```
[the remark this answers is not in the document]
Comment (dzolt-4chain) [reply engine/contact/testabilities/assert_contact.go:48]: oh you're right :D

File: engine/contact/contact_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:59]: sure :D

File: engine/contact/contact_test.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+		expectedFullName string
+	}{
+		"Happy path without metadata": {
+			contactPaymail:   fixtures.RecipientExternal.DefaultPaymail(),
+			creatorPaymail:   fixtures.Sender.DefaultPaymail(),
+			fullName:         "John Doe",
+			metadata:         nil,
+			expectedError:    nil,
+			expectedStatus:   engine.ContactNotConfirmed,
+			expectedFullName: "John Doe",
```
[the remark this answers is not in the document]
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:31]: you're right. changed

File: engine/contact/contact_test.go (hunk 3 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:56]: makes sense changed :D

File: engine/contact/contact_test.go (hunk 6 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:92]: I didn't change original names :D Will change them

File: engine/contact/testabilities/assert_contact.go (hunk 10 above)
Comment (dzolt-4chain) [reply engine/contact/testabilities/assert_contact.go:61]: I would leave it as is, because I believe in the end we will want to move to testabilities in the whole code base and there are other methods that test different statuses like [here](https://github.com/bitcoin-sv/spv-wallet/blob/main/engine/contact_service_test.go#L134) and [here](https://github.com/bitcoin-sv/spv-wallet/blob/main/engine/contact_service_test.go#L213)

File: engine/contact/contact_test.go (hunk 12 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:90]: removed

File: engine/contact/contact_test.go (hunk 7 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:111]: added

File: engine/contact/contact_test.go (hunk 4 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go]: done 🚀

File: engine/contact/testabilities/assert_contact.go (hunk 9 above)
Comment (dzolt-4chain) [reply engine/contact/testabilities/assert_contact.go]: changed to assert

File: engine/contact/contact_test.go (hunk 5 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go]: done :D

**Review by chris-4chain — APPROVED** [review]

File: engine/contact/contact_test.go [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+			contact, err := service.AdminCreateContact(context.Background(),
+				tt.contactPaymail,
+				tt.creatorPaymail,
+				tt.fullName,
+				tt.metadata,
+			)
+
+			//then:
+			then.
+				Contact(contact).
```
Comment (dorzepowski) [inline engine/contact/contact_test.go:56]: ### minor
```suggestion
				Created(contact).
```

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/contact/contact_test.go (hunk 16 above)
Comment (dzolt-4chain) [reply engine/contact/contact_test.go:56]: To me it seams clearer when its Contact but to be consistent with previous implementations I'm changing it here

File: engine/contact/testabilities/assert_contact.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+func (a *assertion) WithNoError(err error) ContactSuccessAssertion {
+	a.require.NotNil(a.contact, "unexpected nil response")
+	a.require.NoError(err, "record transaction outline has error")
+	return a
+}
+
+func (a *assertion) ThatIs(expectedError error) ContactFailureAssertion {
+	a.require.Nil(a.contact, "unexpected response")
```
[the remark this answers is not in the document]
Comment (dzolt-4chain) [reply engine/contact/testabilities/assert_contact.go:63]: yeah, also just noticed it thanks for pointing it out :D

**Review by dorzepowski — APPROVED** [review]
